libs/base/Data/Vect/Elem.idr

```diff
@@ -0,0 +1,73 @@
+module Data.Vect.Elem
+
+import Data.Vect
+import Decidable.Equality
+
+--------------------------------------------------------------------------------
+-- Vector membership proof
+--------------------------------------------------------------------------------
+
+||| A proof that some element is found in a vector
+public export
+data Elem : a -> Vect k a -> Type where
+  Here : Elem x (x::xs)
+  There : (later : Elem x xs) -> Elem x (y::xs)
+
+export
+Uninhabited (Elem x []) where
+  uninhabited Here impossible
+
+||| An item not in the head and not in the tail is not in the Vect at all
+export
+neitherHereNorThere : {x, y : a} -> {xs : Vect n a} -> Not (x = y) -> Not (Elem x xs) -> Not (Elem x (y :: xs))
+neitherHereNorThere xneqy xninxs  Here         = xneqy Refl
+neitherHereNorThere xneqy xninxs (There xinxs) = xninxs xinxs
+
+||| A decision procedure for Elem
+public export
+isElem : DecEq a => (x : a) -> (xs : Vect n a) -> Dec (Elem x xs)
+isElem x [] = No uninhabited
+isElem x (y::xs) with (decEq x y)
+  isElem x (x::xs) | (Yes Refl) = Yes Here
+  isElem x (y::xs) | (No xneqy) with (isElem x xs)
+    isElem x (y::xs) | (No xneqy) | (Yes xinxs) = Yes (There xinxs)
+    isElem x (y::xs) | (No xneqy) | (No xninxs) = No (neitherHereNorThere xneqy xninxs)
+
+public export
+replaceElem : (xs : Vect k t) -> Elem x xs -> (y : t) -> (ys : Vect k t ** Elem y ys)
+replaceElem (x::xs) Here y = (y :: xs ** Here)
+replaceElem (x::xs) (There xinxs) y with (replaceElem xs xinxs y)
+  replaceElem (x::xs) (There xinxs) y | (ys ** yinys) = (x :: ys ** There yinys)
+
+public export
+replaceByElem : (xs : Vect k t) -> Elem x xs -> t -> Vect k t
+replaceByElem (x::xs)  Here         y = y :: xs
+replaceByElem (x::xs) (There xinxs) y = x :: replaceByElem xs xinxs y
+
+public export
+mapElem : {0 xs : Vect k t} -> {0 f : t -> u} ->
+          Elem x xs -> Elem (f x) (map f xs)
```

Comment:
I expect we can make it `(1 _ : Elem x xs)`.
Not sure whether we should always be as precise as possible in the lib?

Reply:
Let's try, I guess :)